File: utils/rate_limiter.py
# -*- coding: utf-8 -*-
"""
@file rate_limiter.py
@description Lightweight in-memory rate limiting implementation for Python Flask app.
"""

import time
from flask import request, jsonify

# Store rate limiting logs in-memory
ip_records = {}

def clean_expired_records(window_sec=60):
    """
    Cleans up old rate limit hits.
    """
    now = time.time()
    for ip in list(ip_records.keys()):
        hits = [t for t in ip_records[ip] if now - t < window_sec]
        if not hits:
            del ip_records[ip]
        else:
            ip_records[ip] = hits

def limit_requests(limit=10, window_sec=60):
    """
    Flask decorator / interceptor to rate limit requests.
    """
    def decorator(f):
        def wrapper(*args, **kwargs):
            # Safe IP resolution
            ip = request.headers.get("X-Forwarded-For", request.remote_addr or "127.0.0.1")
            ip = ip.split(",")[0].strip()
            
            now = time.time()
            clean_expired_records(window_sec)
            
            if ip not in ip_records:
                ip_records[ip] = []
                
            # Filter hits outside the window
            ip_records[ip] = [t for t in ip_records[ip] if now - t < window_sec]
            
            if len(ip_records[ip]) >= limit:
                return jsonify({
                    "response": "⚠️ **Rate Limit Exceeded**: You are posting inputs too fast! Please pause and try again in a minute.",
                    "detected_language": "English"
                }), 429
                
            ip_records[ip].append(now)
            return f(*args, **kwargs)
        wrapper.__name__ = f.__name__
        return wrapper
    return decorator

File: server-security/rate_limiter.ts
/**
 * @file rate_limiter.ts
 * @description Bulletproof in-memory rate limiter middleware for Express to prevent DDoS and API abuse.
 */

import { Request, Response, NextFunction } from "express";

interface RateLimitRecord {
  timestamps: number[];
}

// Memory-efficient IP request log mapping
const ipLimits = new Map<string, RateLimitRecord>();

// Clean up memory maps periodically to prevent memory leaks
setInterval(() => {
  const now = Date.now();
  for (const [ip, record] of ipLimits.entries()) {
    // Retain only requests within the last minute (60,000 ms)
    record.timestamps = record.timestamps.filter(ts => now - ts < 60000);
    if (record.timestamps.length === 0) {
      ipLimits.delete(ip);
    }
  }
}, 300000); // Trigger every 5 minutes

/**
 * Express middleware to enforce a request limit per IP.
 * Defaults to 10 requests per minute.
 */
export function rateLimiter(limit = 10, windowMs = 60000) {
  return (req: Request, res: Response, next: NextFunction) => {
    // Resolve requester IP securely
    const rawIp = req.headers["x-forwarded-for"] || req.socket.remoteAddress || "127.0.0.1";
    const ip = (Array.isArray(rawIp) ? rawIp[0] : typeof rawIp === "string" ? rawIp.split(",")[0] : "127.0.0.1").trim();
    
    const now = Date.now();
    let record = ipLimits.get(ip);
    
    if (!record) {
      record = { timestamps: [] };
      ipLimits.set(ip, record);
    }
    
    // Filter timestamps falling outside the sliding window
    record.timestamps = record.timestamps.filter(ts => now - ts < windowMs);
    
    if (record.timestamps.length >= limit) {
      console.warn(`[SECURITY - RATE LIMIT] Limit exceeded for IP: ${ip} on path: ${req.path}`);
      return res.status(429).json({
        response: "⚠️ **Rate Limit Exceeded**: You are sending messages too quickly! To maintain high system availability, please wait a minute before trying again.",
        detected_language: "English"
      });
    }
    
    // Log the current valid hit
    record.timestamps.push(now);
    next();
  };
}
